File: leetcode/rust/src/50.powx-n.rs
impl Solution {
    // 快速幂运算
    pub fn my_pow(x: f64, n: i32) -> f64 {
        if n == 0 {
            return 1.0;
        }

        let is_neg :bool = n < 0;

        let mut abs_n = n;
        if abs_n < 0{
            abs_n = -abs_n;
        }
        let mut res = Self::my_pow_inner(x,n);

        if is_neg {
            return 1.0f64 / res;
        }
        res

    }

    // n >= 0
    pub fn my_pow_inner(x : f64,n:i32) -> f64{
        if n == 0{
            return 1.0;
        }
        
        if n == 1{
            return x;
        }

        let mut res = Self::my_pow_inner(x,n/2);
        res *= res;

        if (n & 1) == 1{
            res *= x;
        }
        res

    }
}

File: leetcode/java/50.powx-n.java
// 快速幂运算
class Solution {
    public double myPow(double x, int n) {
        boolean isNeg = n < 0;
        if (n == 0)
            return 1;

        long m = Math.abs(n);

        double res = pow(x, m);

        if (isNeg) {
            return 1d / res;
        }
        return res;

    }

    private double pow(double base, long n) {
        if (n == 0)
            return 1;
        if (n == 1)
            return base;

        long half = n / 2;
        double part = pow(base, half);
        part *= part;
        if ((n & 1) == 1) {
            part *= base;
        }

        return part;
    }
}
